base_layer/core/src/base_node/states/block_sync.rs

```diff
@@ -67,213 +96,309 @@ impl BlockSyncInfo {
-) -> Result<StateEvent, String>
+) -> Result<(), BlockSyncError>
 {
-    let local_metadata = shared.db.get_metadata().map_err(|e| e.to_string())?;
-    let mut selected_sync_peer = select_sync_peer(sync_peers);
-
-    if let Some(mut sync_block_hash) = network_metadata.best_block.clone() {
-        // Find the missing block hashes of the strongest network chain.
-        let mut attempts: usize = 0;
-        let mut block_hashes = VecDeque::<BlockHash>::new();
-        let mut linked_to_chain = false;
-        while local_metadata.accumulated_difficulty.unwrap_or_else(|| 0.into()) <
-            network_metadata.accumulated_difficulty.unwrap_or_else(|| 0.into())
+    let local_metadata = db.get_metadata()?;
+    if let Some(local_block_hash) = local_metadata.best_block.clone() {
+        info!(
+            target: LOG_TARGET,
+            "Checking if current chain lagging on best network chain."
+        );
+        let local_tip_height = local_metadata.height_of_longest_chain.unwrap_or(0);
+        let network_tip_height = network_metadata.height_of_longest_chain.unwrap_or(0);
+        let mut sync_height = local_tip_height + 1;
+        if check_chain_split(
+            comms,
+            config,
+            sync_peers,
+            local_tip_height,
+            network_tip_height,
+            &local_block_hash,
+        )
+        .await?
         {
-            // Check if sync hash is on local chain.
-            if async_db::fetch_header_with_block_hash(shared.db.clone(), sync_block_hash.clone())
-                .await
-                .is_ok()
+            trace!(target: LOG_TARGET, "Chain split detected, finding chain split height.");
+            let min_tip_height = min(local_tip_height, network_tip_height);
+            sync_height = find_chain_split_height(db, comms, config, sync_peers, min_tip_height).await?;
+            trace!(target: LOG_TARGET, "Chain split found at height {}.", sync_height);
+        } else {
+            trace!(
+                target: LOG_TARGET,
+                "Block hash {} is common between our chain and the network.",
+                local_block_hash.to_hex()
+            );
+        }
+
+        info!(target: LOG_TARGET, "Synchronize missing blocks.");
+        for height in sync_height..=network_tip_height {
+            request_and_add_block(db, comms, config, sync_peers, height).await?;
+        }
+        return Ok(());
+    }
+    Err(BlockSyncError::EmptyBlockchain)
+}
+
+// Perform a basic check to determine if a chain split has occurred between the local and network chain.
+async fn check_chain_split(
+    comms: &mut OutboundNodeCommsInterface,
+    config: &BlockSyncConfig,
+    sync_peers: &[NodeId],
+    local_tip_height: u64,
+    network_tip_height: u64,
+    local_block_hash: &BlockHash,
+) -> Result<bool, BlockSyncError>
+{
+    Ok(if network_tip_height >= local_tip_height {
+        let header = request_header(comms, config, sync_peers, local_tip_height).await?;
```

Comment:
You could possibly save one network request if the tip heights are equal, because the best block hash is already provided in chain_meta_data, right?

Reply:
Yes, you are right that is a good optimisation.  The function was changed to check when they are equal and then use the provided network metadata.